crates/providers/src/lib.rs

```diff
@@ -134,22 +355,201 @@ impl<N: Network, T: Transport + Clone> Provider<N, T> for NetworkRpcClient<N, T>
         panic!("called inner on <RpcClient as Provider>")
     }
 
-    async fn estimate_gas(
+    async fn get_transaction_count(
         &self,
-        tx: &<N as Network>::TransactionRequest,
+        address: Address,
+        tag: Option<BlockId>,
     ) -> TransportResult<alloy_primitives::U256> {
-        self.prepare("eth_estimateGas", Cow::Borrowed(tx)).await
+        self.client.prepare("eth_getTransactionCount", (address, tag.unwrap_or_default())).await
     }
 
-    async fn get_transaction_count(
+    async fn get_block_number(&self) -> TransportResult<u64> {
+        self.client.prepare("eth_blockNumber", ()).await.map(|num: U64| num.to::<u64>())
+    }
+
+    async fn get_balance(&self, address: Address, tag: Option<BlockId>) -> TransportResult<U256> {
+        self.client
+            .prepare(
+                "eth_getBalance",
+                (address, tag.unwrap_or(BlockId::Number(BlockNumberOrTag::Latest))),
+            )
+            .await
+    }
+
+    async fn get_block_by_hash(
+        &self,
+        hash: BlockHash,
+        full: bool,
+    ) -> TransportResult<Option<Block>> {
+        self.client.prepare("eth_getBlockByHash", (hash, full)).await
+    }
+
+    async fn get_block_by_number(
+        &self,
+        number: BlockNumberOrTag,
+        full: bool,
+    ) -> TransportResult<Option<Block>> {
+        self.client.prepare("eth_getBlockByNumber", (number, full)).await
+    }
+
+    async fn get_client_version(&self) -> TransportResult<String> {
+        self.client.prepare("web3_clientVersion", ()).await
+    }
+
+    async fn get_chain_id(&self) -> TransportResult<U64> {
+        self.client.prepare("eth_chainId", ()).await
+    }
+
+    async fn get_net_version(&self) -> TransportResult<U64> {
+        self.client.prepare("net_version", ()).await
+    }
+
+    async fn get_storage_at(
         &self,
         address: Address,
-    ) -> TransportResult<alloy_primitives::U256> {
-        self.prepare(
-            "eth_getTransactionCount",
-            Cow::<(Address, String)>::Owned((address, "latest".to_string())),
-        )
-        .await
+        key: U256,
+        tag: Option<BlockId>,
+    ) -> TransportResult<StorageValue> {
+        self.client.prepare("eth_getStorageAt", (address, key, tag.unwrap_or_default())).await
```

Comment:
generally find it weird we're calling this `self.client.prepare` when it's sending the request, shouldnt it be `self.client.prepare(...).send().await` or `self.client.request(...)`?

Reply:
I think it's called `prepare` because it doesn't do any serialization or really anything at all until `await`ed. The `send` method in your example is awaiting it